Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm

```diff
@@ -1428,15 +1430,37 @@ - (void)_runningOrTerminatingServiceWorkerCountForTesting:(void(^)(NSUInteger))c
     });
 }
 
-- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(WK_SWIFT_UI_ACTOR void (^)(NSData *))completionHandler
+enum class WKWebsiteDataType : uint8_t {
+    WKWebsiteDataTypeLocalStorage
+};
+
+using WKWebsiteDataTypeSet = HashSet<WKWebsiteDataType, IntHash<WKWebsiteDataType>, WTF::StrongEnumHashTraits<WKWebsiteDataType>>;
+
+namespace WTF {
+
+template<> struct EnumTraitsForPersistence<WKWebsiteDataType> {
+    using values = EnumValues<
+        WKWebsiteDataType,
+        WKWebsiteDataType::WKWebsiteDataTypeLocalStorage
+    >;
+};
+
+}
+
+- (void)_fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(void(^)(NSData *))completionHandler
 {
     if ([dataTypes containsObject:WKWebsiteDataTypeLocalStorage]) {
         _websiteDataStore->fetchLocalStorage([completionHandler = makeBlockPtr(completionHandler)](HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorage) {
             constexpr unsigned currentLocalStorageSerializationVersion = 1;
 
             WTF::Persistence::Encoder encoder;
             encoder << currentLocalStorageSerializationVersion;
-            encoder << localStorage;
+            if (!localStorage.isEmpty()) {
+                encoder << WKWebsiteDataTypeSet { WKWebsiteDataType::WKWebsiteDataTypeLocalStorage };
```

Comment:
The set should be encoded as a pre-step and include all the types included in `dataTypes` and that we support restoration for.